### PR TITLE
feature: Support "jws" Linked Data proof

### LINKED ISSUE #1093: Support "jws" Linked Data proof
Currently, we support the verification of Linked Data proofs which takes value to verify from `proofValue` element of the `proof` of JSON-LD document (e.g. Verifiable Credential). For example:
```
"proof": {
  "type": "Ed25519Signature2018",
  "created": "2020-01-21T16:44:53+02:00",
  "proofValue": "eyJhbGciOiJSUzI1NiIsImI2NCI6ZmFsc2UsImNyaXQiOlsiYjY0Il19..kTCYt5XsITJX1CxPCT8yAV-TVIw5WEuts01mq-pQy7UJiN5mgREEMGlv50aqzpqh4Qq_PbChOMqsLfRoPsnsgxD-WUcX16dUOqV0G_zS245-kronKb78cPktb3rk-BuQy72IFLN25DYuNzVBAh4vGHSrQyHUGlcTwLtjPAnKb78"
}
```

Next, we need to support proof values kept in `jws` element, for example:
```
"proof": {
  "type": "Ed25519Signature2018",
  "created": "2020-01-20T16:11:36Z",
  "jws": "eyJhbGciOiJFZERTQSIsImI2NCI6ZmFsc2UsImNyaXQiOlsiYjY0Il19..67TTULBvibJaJ2oZf3tGYhxZqxYS89qGQykL5hfCoh-MF0vrwQqzciZhjNrAGTAgHtDZsnSQVwJ8bO_7Sc0ECw",
  "proofPurpose": "assertionMethod"
}
```

The reference implementation could be https://github.com/digitalbazaar/jsonld-signatures

To be used by Verifiable Credentials.

### PR DESCRIPTION
**Pull request by kdimak**
closes #1093

Signed-off-by: Dima <dkinoshenko@gmail.com>

Based on https://github.com/digitalbazaar/jsonld-signatures

### CODE REVIEW

File: pkg/doc/signature/proof/data.go [hunk 1]
```diff
@@ -19,6 +22,32 @@ type signatureSuite interface {
 	GetDigest(doc []byte) []byte
 }
 
+// SignatureHolding defines a representation of signature value.
+type SignatureHolding int
```
Comment (troyronda) [inline pkg/doc/signature/proof/data.go]: I'm not finding "Holding" to be an intuitive description of this type.

Comment (kdimak) [reply pkg/doc/signature/proof/data.go]: It's a sort of "a type of signature value holder". Maybe `SignatureRepresentation` will sound better? Or "SignatureHoldingType"?

Comment (troyronda) [reply pkg/doc/signature/proof/data.go]: probably SignatureRepresentation is better. the word holding confuses me for an enumerated type.

Comment (kdimak) [reply pkg/doc/signature/proof/data.go]: Thank you! renamed.

File: pkg/doc/signature/proof/data.go [hunk 2]
```diff
@@ -19,6 +22,32 @@ type signatureSuite interface {
...
+// In case of "jws",
+func CreateVerifyData(suite signatureSuite, jsonldDoc map[string]interface{}, proof *Proof) ([]byte, error) {
+	switch proof.SignatureRepresentation {
+	case SignatureProofValue:
+		return CreateVerifyHash(suite, jsonldDoc, proof.JSONLdObject())
+	case SignatureJWS:
+		return createVerifyJWS(suite, jsonldDoc, proof)
+	}
+
+	return nil, fmt.Errorf("unsupported signature value holder: %v", proof.SignatureRepresentation)
```
Comment (troyronda) [inline pkg/doc/signature/proof/data.go]: string wasn't updated to representation.

File: pkg/doc/verifiable/linked_data_proof_test.go [hunk 3]
```diff
@@ -41,7 +44,58 @@ func Test_keyResolverAdapter_Resolve(t *testing.T) {
...
+func TestLinkedDataProofVerifier(t *testing.T) {
+	pubKeyStr := base58.Decode("5fYcSLLNTzCzCRTP7dNJiQ4AcUqB9Jf63Y4bPJ6abEFe")
+	pubKey := ed25519.PublicKey(pubKeyStr)
+
+	//nolint:lll
+	vcStr := `
+{
+  "@context": [
+    "https://www.w3.org/2018/credentials/v1",
+    "https://transmute-industries.github.io/vc-greeting-card/context/vc-greeting-card-v0.0.jsonld"
```
Comment (troyronda) [inline pkg/doc/verifiable/linked_data_proof_test.go]: Will the test break if this URL goes away?

Comment (kdimak) [reply pkg/doc/verifiable/linked_data_proof_test.go]: Yes...

Comment (kdimak) [reply pkg/doc/verifiable/linked_data_proof_test.go]: Fixed!

File: pkg/doc/signature/proof/data.go (hunk 2 above)
Comment (kdimak) [reply pkg/doc/signature/proof/data.go]: fixed

File: pkg/doc/signature/proof/data.go [hunk 4]
```diff
@@ -19,6 +22,32 @@ type signatureSuite interface {
...
+	SignatureProofValue SignatureRepresentation = iota
+
+	// SignatureJWS uses "jws" field in a Proof as an element for representation of detached JSON Web Signatures.
+	SignatureJWS
+)
+
+// CreateVerifyData creates data that is used to generate or verify a digital signature.
+// It depends on the signature value holder type.
+// In case of "proofValue", the standard Create Verify Hash algorithm is used.
+// In case of "jws",
```
Comment (sandrask) [inline pkg/doc/signature/proof/data.go]: unfinished sentence

File: pkg/doc/signature/proof/jws.go [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+	if len(jwtParts) != 3 || jwtParts[2] == "" {
+		return nil, errors.New("invalid JWT")
+	}
+
+	return base64.RawURLEncoding.DecodeString(jwtParts[2])
+}
+
+func getJWTHeader(jwt string) (string, error) {
+	jwtParts := strings.Split(jwt, ".")
+	if len(jwtParts) != 3 { // nolint:gomnd
```
Comment (sandrask) [inline pkg/doc/signature/proof/jws.go]: define 3 as constant e.g. jwtPartNumber instead of using magic numbers

File: pkg/doc/signature/proof/jws.go [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+		return "", errors.New("invalid JWT")
+	}
+
+	return jwtParts[0], nil
```
Comment (sandrask) [inline pkg/doc/signature/proof/jws.go]: also create constants for token parts such as 0 should be jwtHeaderPart, 2 should be jwtSignaturePart etc.

File: pkg/doc/signature/proof/jws.go [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+}
+
+// createVerifyJWS creates a data to be used to create/verify a digital signature.
+// By the time of development, no specification was found with the algorithm description.
+// Current implementation is based on the https://github.com/digitalbazaar/jsonld-signatures.
```
Comment (sandrask) [inline pkg/doc/signature/proof/jws.go]: Did you have a chance to look at this spec: https://tools.ietf.org/html/rfc7797?

File: pkg/doc/signature/proof/data.go [hunk 8]
```diff
@@ -19,6 +22,32 @@ type signatureSuite interface {
...
+func CreateVerifyData(suite signatureSuite, jsonldDoc map[string]interface{}, proof *Proof) ([]byte, error) {
```
Comment (sandrask) [inline pkg/doc/signature/proof/data.go]: CreateVerifyData and CreateVerifyHash have same meaning to create data for signature (creating this data involves hashing thus CreateVerifyHash). We should probably have only one function,  let's discuss.

File: pkg/doc/signature/proof/jws.go (hunk 7 above)
Comment (kdimak) [reply pkg/doc/signature/proof/jws.go]: Yes. I will add a reference to this spec. At https://github.com/digitalbazaar/jsonld-signatures, they use several heuristics which I did not found in any spec:
1) Use https://w3id.org/security/v2 as a `@context` for JSON-LD canonization of both JSON and Signature documents.
2) Apply preliminary JSON-LD compacting of JSON document using the same https://w3id.org/security/v2 `@context`.
3) Remove only `jws` and `proofValue` elements only from the proof document before canonization. At `signature` package (and what is stated in https://w3c-dvcg.github.io/ld-signatures/#create-verify-hash-algorithm), we remove `id`, `type`, `proofValue` and `jws` elements.

File: pkg/doc/signature/proof/data.go (hunk 8 above)
Comment (kdimak) [reply pkg/doc/signature/proof/data.go]: Yes, agree, it makes sense. The reason I added a separate method is that CreateVerifyHash is different in some steps in the case of JWS. I described it above https://github.com/hyperledger/aries-framework-go/pull/1137#discussion_r374120358.

File: pkg/doc/verifiable/linked_data_proof_test.go [hunk 9]
```diff
@@ -11,8 +11,14 @@ import (
 	"crypto/rsa"
 	"errors"
 	"testing"
+	"time"
 
+	"github.com/btcsuite/btcutil/base58"
 	"github.com/stretchr/testify/require"
+
+	"github.com/hyperledger/aries-framework-go/pkg/doc/signature/ed25519signature2018"
+
```
Comment (soluchok) [inline pkg/doc/verifiable/linked_data_proof_test.go]: extra line

File: pkg/doc/verifiable/example_credential_test.go [hunk 10]
```diff
@@ -342,11 +342,11 @@ func ExampleCredential_AddLinkedDataProof() {
 	}
 
 	err = vc.AddLinkedDataProof(&verifiable.LinkedDataProofContext{
-		Created:       &issued,
-		Creator:       "Example University",
-		SignatureType: "Ed25519Signature2018",
-		Suite:         ed25519signature2018.New(),
-		PrivateKey:    privIssuerKey,
+		Created:                 &issued,
+		SignatureType:           "Ed25519Signature2018",
```
Comment (soluchok) [inline pkg/doc/verifiable/example_credential_test.go]: better to create a constant for `"Ed25519Signature2018"`

File: pkg/doc/signature/proof/jws.go [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+
+// CreateDetachedJWTHeader creates detached JWT header.
+func CreateDetachedJWTHeader(p *Proof) string {
+	jwtHeaderMap := map[string]interface{}{
+		"alg":  p.Type,
+		"b64":  false,
+		"crit": []string{"b64"},
+	}
+
+	jwtHeaderBytes, _ := json.Marshal(jwtHeaderMap) //nolint:errcheck
```
Comment (soluchok) [inline pkg/doc/signature/proof/jws.go]: If you want to ignore error handling, better to use panic.

```
jwtHeaderBytes, err := json.Marshal(jwtHeaderMap)
if err != nil {
    panic(err)
}
```

File: pkg/doc/verifiable/example_credential_test.go (hunk 10 above)
Comment (kdimak) [reply pkg/doc/verifiable/example_credential_test.go]: I think it's OK to use the constant as is in the test.

File: pkg/doc/signature/proof/jws.go (hunk 11 above)
Comment (kdimak) [reply pkg/doc/signature/proof/jws.go]: I remember @troyronda advised not to **panic** in production code. @troyronda what's the best approach here? `jwtHeaderMap` is a constant JSON document defined in the same file (`jws.go`).

Comment (soluchok) [reply pkg/doc/signature/proof/jws.go]: In this case, panic is better. At least you can see the problem if it happens.

File: pkg/doc/signature/ed25519signature2018/suite.go [hunk 12]
```diff
@@ -52,7 +52,7 @@ func (s *SignatureSuite) GetCanonicalDocument(doc map[string]interface{}) ([]byt
 
 // GetDigest returns document digest
 func (s *SignatureSuite) GetDigest(doc []byte) []byte {
-	digest := sha512.Sum512(doc)
+	digest := sha256.Sum256(doc)
```
Comment (llorllale) [inline pkg/doc/signature/ed25519signature2018/suite.go]: What is the reason for this change? Where is the test that validates this change?

File: pkg/doc/signature/proof/jws.go (hunk 11 above)
Comment (troyronda) [reply pkg/doc/signature/proof/jws.go]: @kdimak It is true that as a library we must avoid panic.

- https://github.com/golang/go/wiki/CodeReviewComments#handle-errors
- https://golang.org/doc/effective_go.html#panic
- https://github.com/uber-go/guide/blob/master/style.md#dont-panic

For error situations that are supposed to be impossible, this is an exceptional situation - something must be seriously wrong with the execution environment (or something seriously wrong with an update to some library or code we are using).

File: pkg/doc/signature/proof/jws.go (hunk 5 above)
Comment (troyronda) [reply pkg/doc/signature/proof/jws.go]: Agree- shouldn't nolint this one.

File: pkg/doc/signature/ed25519signature2018/suite.go (hunk 12 above)
Comment (kdimak) [reply pkg/doc/signature/ed25519signature2018/suite.go]: Please see https://w3c-dvcg.github.io/ld-signatures/#create-verify-hash-algorithm, ISSUE 8:
> The Ed25519Signature2018 algorithm also does not perform this last step -- and, in fact, uses SHA-512 internally. In short, this last step should better communicate that the 64 bytes produced from concatenating the SHA-256 of the canonicalized options with the SHA-256 of the canonicalized document are passed into the signing algorithm with a presumption that the signing algorithm will include hashing of its own.

We will call GetDigest() twice and concatenate the result, so it's worth to have it 32+32 bytes.

Comment (kdimak) [reply pkg/doc/signature/ed25519signature2018/suite.go]: Here is how it's implemented in https://github.com/digitalbazaar/jsonld-signatures:

https://github.com/digitalbazaar/jsonld-signatures/blob/f1e78c2674a786c3cdc8d310dbf228ed8a1e1b8e/lib/suites/LinkedDataSignature.js#L213

File: pkg/doc/signature/proof/jws.go (hunk 11 above)
Comment (kdimak) [reply pkg/doc/signature/proof/jws.go]: @troyronda thank you for the really detailed explanation! So I will put a panic() here as @soluchok suggested, thanks Andrew!

File: pkg/doc/verifiable/linked_data_proof_test.go (hunk 9 above)
Comment (kdimak) [reply pkg/doc/verifiable/linked_data_proof_test.go]: Fixed

File: pkg/doc/signature/proof/jws.go (hunk 11 above)
Comment (kdimak) [reply pkg/doc/signature/proof/jws.go]: Corrected.

File: pkg/doc/signature/proof/data.go (hunk 4 above)
Comment (kdimak) [reply pkg/doc/signature/proof/data.go]: fixed

File: pkg/doc/signature/proof/jws.go (hunk 5 above)
Comment (kdimak) [reply pkg/doc/signature/proof/jws.go]: fixed

File: pkg/doc/signature/proof/jws.go (hunk 6 above)
Comment (kdimak) [reply pkg/doc/signature/proof/jws.go]: fixed

File: pkg/doc/signature/proof/data.go [hunk 13]
```diff
@@ -86,14 +110,15 @@ const (
 	proofType excludedKey = iota + 1
 	proofID
 	proofValue
+	jws
 )
 
 func (ek excludedKey) String() string {
-	return [...]string{"type", "id", "proofValue"}[ek-1]
+	return [...]string{"type", "id", "proofValue", "jws"}[ek-1]
```
Comment (soluchok) [inline pkg/doc/signature/proof/data.go]: I would make it a global vav  [...]string{"type", "id", "proofValue", "jws"}

File: pkg/doc/signature/proof/data.go [hunk 14]
```diff
@@ -86,14 +110,15 @@ const (
...
 }
 
 func excludedKeyFromString(s string) excludedKey {
-	for _, ek := range [...]excludedKey{proofType, proofID, proofValue} {
+	for _, ek := range [...]excludedKey{proofType, proofID, proofValue, jws} {
```
Comment (soluchok) [inline pkg/doc/signature/proof/data.go]: And this one as well under the const definition `[...]excludedKey{proofType, proofID, proofValue, jws} `

File: pkg/doc/signature/proof/jws.go [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+		return nil, err
+	}
+
+	return append([]byte(jwtHeader+"."), verifyData...), nil
+}
+
+func prepareJWSProof(suite signatureSuite, proofOptions map[string]interface{}) ([]byte, error) {
+	proofOptions[jsonldContext] = securityContext
+
+	proofOptionsCopy := make(map[string]interface{})
```
Comment (soluchok) [inline pkg/doc/signature/proof/jws.go]: Capacity can be added e.g make(map[string]interface{},len(proofOptions))

File: pkg/doc/verifiable/linked_data_proof_test.go [hunk 16]
```diff
@@ -41,7 +46,107 @@ func Test_keyResolverAdapter_Resolve(t *testing.T) {
...
 	})
 }
+
+type dummyKeyResolver struct {
+	pubKey []byte
+}
+
+func (kr *dummyKeyResolver) Resolve(string) ([]byte, error) {
+	return kr.pubKey, nil
+}
```
Comment (soluchok) [inline pkg/doc/verifiable/linked_data_proof_test.go]: Should be cheaper and simpler e.g usage `dummyKeyResolver(bytes)`
```
type dummyKeyResolver []byte

func (kr dummyKeyResolver) Resolve(string) ([]byte, error) {
	return kr, nil
}
```

File: pkg/doc/signature/proof/data.go (hunk 13 above)
Comment (kdimak) [reply pkg/doc/signature/proof/data.go]: It's a matter of taste, I think. But in this particular case, I will create two constants for `[...]string{"type", "id", "proofValue"}` and `[...]excludedKey{proofType, proofID, proofValue, jws} ` so that they will be placed together and it will be easier to refine them in future.

File: pkg/doc/verifiable/linked_data_proof_test.go (hunk 16 above)
Comment (kdimak) [reply pkg/doc/verifiable/linked_data_proof_test.go]: Yes, I agree with that. On one hand, it's a struct definition in test code. But we need to strive to have test code even better than production one. :)
